### PR TITLE
Revolving credit

### PR DESCRIPTION
**Pull request by davidyaha**

### CODE REVIEW

File: fineract-provider/src/main/resources/sql/migrations/core_db/V366__add_revolving_period_support.sql [hunk 1]
```diff
@@ -0,0 +1,25 @@
+--
```
Comment (eitanfr) [inline fineract-provider/src/main/resources/sql/migrations/core_db/V366__add_revolving_period_support.sql:1]: sync with @tomermoshe  about the number

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanApiConstants.java [hunk 2]
```diff
@@ -22,6 +22,8 @@
 
     String emiAmountParameterName = "fixedEmiAmount";
     String maxOutstandingBalanceParameterName = "maxOutstandingLoanBalance";
+    String revolvingPeriodStartParameterName = "revolvingPeriodStartDate";
+    String revolvingPeriodEndParameterName = "revolvingPeriodEndDate";
```
Comment (eitanfr) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanApiConstants.java:26]: add your changes to api doc html

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 3]
```diff
@@ -518,6 +531,16 @@ private Loan(final String accountNo, final Client client, final Group group, fin
         // rates added here
         this.rates = rates;
 
+        if (loanProduct.isRevolving()) {
+            if (revolvingPeriodStartDate != null) {
```
Comment (eitanfr) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:542]: why `!=` null?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 4]
```diff
@@ -518,6 +531,16 @@ private Loan(final String accountNo, final Client client, final Group group, fin
...
+                this.revolvingPeriodStartDate = Date
+                        .from(revolvingPeriodStartDate.atStartOfDay(DateUtils.getDateTimeZoneOfTenant()).toInstant());
+                if (revolvingPeriodEndDate != null) {
+                    this.revolvingPeriodEndDate = Date
+                            .from(revolvingPeriodEndDate.atStartOfDay(DateUtils.getDateTimeZoneOfTenant()).toInstant());
```
Comment (eitanfr) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: change to `endOfDay`

File: fineract-provider/src/main/resources/sql/migrations/core_db/V366__add_revolving_period_support.sql (hunk 1 above)
Comment (davidyaha) [reply fineract-provider/src/main/resources/sql/migrations/core_db/V366__add_revolving_period_support.sql:1]: done

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 3 above)
Comment (davidyaha) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:542]: I leave the option to not define the revolving period, and it should mean that the credit can always revolve

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java [hunk 5]
```diff
@@ -2727,6 +2727,8 @@ private CommandProcessingResult processLoanDisbursementDetail(final Loan loan, L
             }
         }
 
+        loan.updateLoanTransactionMappings();
```
Comment (eitanfr) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java:2751]: say, what about the code that does the copy for the transactions, maybe it is a better place to add this fix?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanApiConstants.java [hunk 6]
```diff
@@ -22,6 +22,8 @@
...
     String disbursementDataParameterName = "disbursementData";
     String disbursementDateParameterName = "expectedDisbursementDate";
```
Comment (eitanfr) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanApiConstants.java:28]: do you expose all the new fields for the client?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 4 above)
Comment (davidyaha) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: there is no such method... regardless, this should be used without accounting for the time... I think...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProductTrancheDetails.java [hunk 7]
```diff
@@ -55,6 +59,15 @@ public void update(final JsonCommand command, final Map<String, Object> actualCh
             this.multiDisburseLoan = newValue;
         }
 
+        if (command.isChangeInBooleanParameterNamed(LoanProductConstants.REVOLVING_LOAN_PARAMETER_NAME, this.isRevolving)) {
+            final boolean newValue = command.booleanPrimitiveValueOfParameterNamed(LoanProductConstants.MULTI_DISBURSE_LOAN_PARAMETER_NAME);
+            actualChanges.put(LoanProductConstants.REVOLVING_LOAN_PARAMETER_NAME, newValue);
+            this.isRevolving = newValue;
+        }
+
+        // XXX TODO: maybe check that there is no way for creating a product that is revolving but don't have
```
Comment (eitanfr) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProductTrancheDetails.java:68]: if we don't set multiDisbursement then the effect of the revolving period will be a range when we can do disbursement (only once). So maybe it is ok not having this kind of check

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java (hunk 5 above)
Comment (davidyaha) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java:2751]: I think regardless of where you use this method, it should still be a method of `Loan`.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanApiConstants.java (hunk 6 above)
Comment (davidyaha) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanApiConstants.java:28]: I don't know. Getting into it now...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProductTrancheDetails.java (hunk 7 above)
Comment (davidyaha) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProductTrancheDetails.java:68]: seems like an unintended effect though...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 4 above)
Comment (eitanfr) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: why accounting is related? 
any why, in the current implementation you cant do a disbursement in the end date which is wrong.

https://stackoverflow.com/questions/36408548/how-to-obtain-the-end-of-the-day-when-given-a-localdate

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java (hunk 5 above)
Comment (eitanfr) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java:2751]: true, but I asked if ut should be called somewhere else...

**Pull request merged**